### PR TITLE
Weekly Mood Insights API

### LINKED ISSUE #73: Weekly Insights History Endpoint (/api/v1/insights/history)
**Goal:** Implement backend endpoint to return 7-day mood history for graph and timeline insights.

**Endpoint:**
GET /api/v1/insights/history?week_offset=N

**Expected Behavior:**

- Returns a 7-day graph (date, mood_value, mood_id)
- Returns timeline (one note per day, latest wins, with snippet)
- Days with no notes are included in graph with null mood
- Uses mood_value_snapshot if present, else mood.category.value
- Supports SQLite
- Response cached per user+week (TTL: 300s)

**Checklist:**

- [ ] Query notes for current user within computed week range
- [ ]  One note per day (latest)
- [ ]  Graph: 7 items
- [ ]  Timeline: only days with notes
- [ ]  Caching with TTL
- [ ]  SQLite-safe ORM logic
- [ ]  Tests: week math, user scoping, missing days
- [ ]  Add sample request/response to API docs

### PR DESCRIPTION
**Pull request by buyekeobare**
**Description:**
Implements the /api/v1/insights/history endpoint 

**Features:**

**- GET /api/v1/insights/history?week_offset=N → returns:**
- [ ]  graph: 7 mood points (Mon–Sun)
- [ ] - timeline: per-day note snippets
**- Supports:**
- [ ] Missing days (graph shows null)
- [ ] Mood snapshot logic (mood_value_snapshot preferred)
- [ ] Efficient query using latest note per day
- [ ] Caching per (user, week_offset) for 5 minutes

**Backend Implementation:**

- insights/views.py: APIView + request validation
- insights/services.py: week range logic, cache, timeline, graph
- notes/signals.py: auto-fill mood_value_snapshot from Mood.category
- notes/migrations/0005_backfill_mood_value_snapshot.py: backfills existing notes
- insights/tests/test_history.py: full test suite (7d graph, user scoping, nulls, cache, snapshot)
- insights/urls.py: added named route insights-history

 **Logic:**

- insights/services.py builds payload using fetch_weekly_latest_per_day() 
- insights/views.py handles request validation
- notes/signals.py auto-fills snapshot via pre_save
- Fallback: maps Mood.category (string) to 1–5 scale
- Caching via Django’s cache framework
- Snapshot backfill migration included

**Checklist:**

- GET /api/v1/insights/history returns graph + timeline
-  Accepts week_offset or week as query param
-  Uses Note.mood_value_snapshot or fallback category
-  Signal auto-fills snapshot on note create/update
-  Backfill migration for historical notes
-  Unit tests for graph, timeline, cache, scoping, null days
-  Caching with TTL = 300s
-  API documented via example and internal docstring

**How to Test:**

GET /api/v1/insights/history?week_offset=0
Authorization: Token <user-token>

**Expect Response:**
Insights Response

- 7 graph points (some may be null)
- timeline with snippets, mood_id, and date

**Other Changes:**

Updated INSTALLED_APPS:
"notes.apps.NotesConfig"

→ Enables Django to load signals.py, ensuring Note.mood_value_snapshot is auto-filled correctly.

**Recent Updates**

**File Updated: `backend/insights/services.py`**

- **Clarified time zone logic** in `get_week_range()` docstring to explicitly state that it uses the active time zone as defined by Django's `TIME_ZONE` setting.
- **Optimized mood query** in `fetch_weekly_latest_per_day()` by adding `select_related("mood")` to eliminate N+1 DB queries.
- **Adjusted `week.end` value** in the `/insights/history` response to be inclusive (i.e., returns Sunday as the final day), improving frontend alignment.
- **Removed duplicated comment header** for cleaner code readability.

**Test Coverage Added**

- Verified `week.end` is inclusive (Sunday) using date assertions.
- Confirmed support for the `?week=` alias parameter (same behavior as `week_offset`).
- Asserted that every graph point includes a `mood_id` field (even if `null`).
- Ensured `mood_value_snapshot` overrides any mood.category-derived fallback.
- Added test to check snippet truncation at 120 characters with ellipsis.

These additions improve confidence in the `/api/v1/insights/history` contract and ensure edge cases are fully covered.

**Related Issue**

Closes #73

### CODE REVIEW

**Review by EgbaFrank — COMMENTED** [review]
Good work, a robust implementation.

File: backend/insights/services.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
+# insights/services.py
+# insights/services.py
+from dataclasses import dataclass
```
[the remark this answers is not in the document]
Comment (buyekeobare) [reply backend/insights/services.py]: Resolved: Removed the duplicate comment header

**Comment by buyekeobare** [comment]
Update: Added a small optimization with select_related("mood") to avoid N+1 queries and adjusted week.end in the response to be inclusive (Sunday). No breaking changes.

File: backend/insights/services.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+class WeekRange:
+    start: date  # inclusive, e.g., Monday 00:00
+    end: date    # exclusive, start + 7 days
+
+
+def get_week_range(week_offset: int) -> WeekRange:
+    """
+    Monday-based weeks in the project's TIME_ZONE.
+    week_offset=0 -> current week; 1 -> previous week; etc.
+    """
```
[the remark this answers is not in the document]
Comment (buyekeobare) [reply backend/insights/services.py]: Resolved: Updated the docstring to clarify that the week range uses the active time zone as defined by Django’s TIME_ZONE setting. Thanks for the suggestion!

**Review by EgbaFrank — COMMENTED** [review]
Good work

**Review by EgbaFrank — APPROVED** [review]
